cpp/src/arrow/io/compressed.cc

```diff
@@ -353,7 +355,10 @@ class CompressedInputStream::Impl {
 
   // Try to feed more data into the decompressed_ buffer.
   Status RefillDecompressed(bool* has_data) {
-    // First try to read data from the decompressor
+    // First try to read data from the decompressor.
+    // This doesn't use `CompressedBufferAvailable()` because when compressed_
+    // exists and available == 0, it might trigger  an empty decompress and set
```

Comment:
It's not necessarily an empty decompress if the decompressor has its own internal buffer?

Reply:
Trivially, if the decompressed data is 1MB of zeros, and we decompress `kDecompressSize` bytes at a time, `DecompressData` will still yield data even after the compressed data is finished.

Reply:
By the way, if `compressed_->size() == 0` here, would it has internal buffer to be decompressed? Feel a little consumed here. If the compressed data length is just 64KB, the first read finished, and when calling `EnsureCompressedData()` again, `compressed_->size() == 0`. Would it require more decompressing here?

Reply:
If the compressed data length is 64 kB, but the decompressed data length is 1 MB, then you'll get several chunks of decompressed data after the compressed data has ended.

Reply:
After re-reading this part of code I finally understand this. Seems CompressedInputStream is far complex than I first thought that...